### PR TITLE
fix: KeepKey produces accurate EIP-712 eth_signTypedData_v4 signatures

### PR DESCRIPTION
**Pull request by 0xMillz**
This PR fixes ethSignTypedData for KeepKey, which was returning inaccurate, unverifiable signatures.

Officially closes the hdwallet part of issue #578 and is fully compatible with EIP-712.

This bug was apparent in issues like [shapeshift/web#5384](https://github.com/shapeshift/web/issues/5384)
after the JSON RPC eth_signTypedData signatures sent to OpenSea were being rejected as invalid.

**Implementation details:**
I went with the `EthereumSignTypedHash` approach for ALL TypedData messages after consulting with @BitHighlander. It seems that the KeepKey firmware is missing the `EthereumTypedDataStructRequest` type, which would be required for a fully human readable, eip-712-compatible message to be displayed on the device screen.

Fortunately, the `EthereumSignTypedHash` approach allows the KeepKey to at least sign the v4 TypedData messages, which for end-users that means being able to do modern web3 interactions like buying and selling NFTs on OpenSea, for example.

-The bulk of the fix is in packages/hdwallet-keepkey/src/ethereum.ts
- I also added a `recoverTypedSignature` check to the sandbox so you know right away if the signature that any wallet spits out is accurate.
-I refactored the Ethereum sandbox tests too. That file is getting way too large and repetitive, but still is the best way to test with KeepKey.
- I couldn't help but fix some other random some bugs along the way.

Thanks!

image

_*KeepKey Bridge didn't seem to be working for me in the sandbox so I used the regular webusb option instead._

### CODE REVIEW

**Comment by BitHighlander** [comment]
This is tested and confirmed to be working. great work Mills

v4 types data is current broken for both shapeshift web and kk desktop users. this is a required pr for addressing and fixing. Issues are demonstrable by using v4 typed data over wallet connect.

👍 :shipit:

**Comment by millsmcilroy** [comment]
> This is tested and confirmed to be working. great work @millsmcilroy
> 
> v4 types data is current broken for both shapeshift web and kk desktop users. this is a required pr for addressing and fixing. Issues are demonstrable by using v4 typed data over wallet connect.
> 
> 👍 :shipit:

ShapeShift team, @0xApotheosis, @gomesalexandre, et al., I'll just standby and wait for a PR review. When it's all ✅ and ready to ship I can update with a version bump or leave that to you. Thx!

File: examples/sandbox/index.ts
Comment (0xMillz) [inline examples/sandbox/index.ts:380]: No reason we have to endure years of this pointless error spamming the console

File: examples/sandbox/index.ts
Comment (0xMillz) [inline examples/sandbox/index.ts:416]: Does not take _any_ args

File: examples/sandbox/index.ts
Comment (0xMillz) [inline examples/sandbox/index.ts:921]: This is the first of many copy/pasta errors with the `$ethResults` etc selector being used for the wrong coin

File: examples/sandbox/index.ts
Comment (0xMillz) [inline examples/sandbox/index.ts:1049]: The TypeScript errors informed me that the fioGetPublicKey method does not exists and it does not

File: examples/sandbox/index.ts
Comment (0xMillz) [inline examples/sandbox/index.ts:2497]: This is a low-risk refactor that ultimately helped me find and fix the bug faster using the powers of TypeScript while also eliminating repetitive code

File: examples/sandbox/index.ts
Comment (0xMillz) [inline examples/sandbox/index.ts:2553]: This change is based on https://github.com/shapeshift/hdwallet/pull/635 that was missed in sandbox. Basically we should be passing a hex string here

File: examples/sandbox/index.ts
Comment (0xMillz) [inline examples/sandbox/index.ts:2602]: Added this utility method to verify the address from the signature. It was previously returning a wrong address before my fix in hdwallet-keepkey/src/ethereum.ts ethSignTypedData. This should help us immediately see any future regressions

File: examples/sandbox/json/ethereum/ethSignTypedDataV4.json
Comment (0xMillz) [inline examples/sandbox/json/ethereum/ethSignTypedDataV4.json:1]: I grabbed this payload straight from the OpenSea request to its own backend server, which was responding with "invalid order signature"

File: examples/sandbox/package.json
Comment (0xMillz) [inline examples/sandbox/package.json:13]: Industry standard library for eth_signTypedData_v4

File: examples/sandbox/tsconfig.json
Comment (0xMillz) [inline examples/sandbox/tsconfig.json:4]: Not sure if we need these changes...was going for less ts-lint errors

File: package.json
Comment (0xMillz) [inline package.json:43]: Recommended change from `yarn`

File: packages/hdwallet-tallyho/src/adapter.ts
Comment (0xMillz) [inline packages/hdwallet-tallyho/src/adapter.ts:26]: Every adapter has this method

File: tsconfig.json
Comment (0xMillz) [inline tsconfig.json:14]: Lmk if this helps anyone's TS IDE intellisense

Comment (kaladinlight) [reply tsconfig.json:14]: Haven't noticed issues and this is pretty much doing the default as far as I can tell. Would probably rather remove at this time

File: package.json
Comment (kaladinlight) [reply package.json:43]: recommended how? not seeing anything related to this and have seen no indication it is needed. hdwallet is already huge, please remove at this time

File: examples/sandbox/tsconfig.json
Comment (kaladinlight) [reply examples/sandbox/tsconfig.json:4]: we can leave it since isolated to sandbox for now though still don't think the `typeRoots` are necessary or making any diff

**Comment by kaladinlight** [comment]
needs to be rebased against master and another package bump. logic looks sane outside of the couple small comments above

File: package.json
Comment (0xMillz) [reply package.json:43]: Hmm I probably meant to add this to sandbox only. It's related to resolving this warning when running `yarn dev:sandbox`, but regardless should probably belong in a different PR anyway
```
▶ yarn dev:sandbox
yarn run v1.22.19
$ cd examples/sandbox && yarn dev
$ yarn parcel index.html
$ /Users/millsmcilroy/workspace/hdwallet/node_modules/.bin/parcel index.html
Server running at http://localhost:1234

console: Browserslist: caniuse-lite is outdated. Please run:
npx browserslist@latest --update-db
Why you should do it regularly: https://github.com/browserslist/browserslist#browsers-data-updating
```

Will remove 👍

Comment (kaladinlight) [reply package.json:43]: ok, yeah that makes more sense. I would say for next time in general, it would be nice to separate the feature logic from the cleanup logic just to make reviews a bit more streamlined. thanks for taking the time though regardless!

Comment (0xMillz) [reply package.json]: Parcel seems to be auto-installing `browserify-zlib` even in CI
Screenshot 2023-10-26 at 20 00 00

Comment (0xMillz) [reply package.json]: No idea why, going to look at this again tomorrow with fresh eyes

File: examples/sandbox/tsconfig.json
Comment (0xMillz) [inline examples/sandbox/tsconfig.json:4]: This makes TS happy when you do `import $ from "jquery"`

Comment (kaladinlight) [reply examples/sandbox/tsconfig.json:4]: yeah this option makes sense :+1:

**Review by kaladinlight — APPROVED** [review]

**Pull request merged**
